File: go/euler30.go
// euler30 находит сумму всех чисел, которые могут быть записаны как
// сумма пятых степеней их цифр.
//
// Решение задачи Эйлера №30.
package main

import (
	"fmt"
	"time"
)

func sumPowerOfDigits(x int, n int) int {
	sum := 0
	for x > 0 {
		digit := x % 10
		pow := 1
		for i := 0; i < n; i++ {
			pow *= digit
		}
		sum += pow
		x /= 10
	}
	return sum
}

func euler30() {
	n := 5

	startTime := time.Now()
	numList := make([]int, 0)
	upperBound := n * powInt(9, n)
	for x := 2; x < upperBound; x++ {
		if sumPowerOfDigits(x, n) == x {
			numList = append(numList, x)
		}
	}
	sum := 0
	for _, v := range numList {
		sum += v
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	sum = 0
	for x := 2; x < upperBound; x++ {
		if sumPowerOfDigits(x, n) == x {
			sum += x
		}
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))
}

func powInt(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

func main() {
	euler30()
}

File: python/euler30.py
# Найдите сумму всех чисел, которые могут быть записаны в виде
# суммы пятых степеней их цифр.

from datetime import datetime


def sum_power_of_digits(x: int, n: int) -> int:
    """
    Возвращает сумму цифр в заданной степени.

    Args:
        x: Число для вычисления.
        n: Степень для возведения цифр.

    Returns:
        Сумма цифр числа в степени n.
    """
    return sum(int(z) ** n for z in str(x))


def euler30() -> None:
    """
    Решение задачи Эйлера №30.

    Найдите сумму всех чисел, которые могут быть записаны в виде
    суммы пятых степеней их цифр.
    """
    n = 5

    # 1 вариант
    start_time = datetime.now()
    num_list = []
    for x in range(2, n * 9 ** n):
        narciss_numbers = sum_power_of_digits(x, n)
        if narciss_numbers == x:
            num_list.append(x)
    print(sum(num_list))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    print(sum(
        i for i in range(2, n * 9 ** n)
        if sum_power_of_digits(i, n) == i
    ))
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler30()
